### ISSUE TITLE
🚀 Epic — Implement radiologist-etl pipeline

### ISSUE THREAD

**Issue by CedrickArmel** [CedrickArmel/radiologist#12]
# 🚀 Epic — Implement radiologist-etl pipeline

Full ETL from raw X-ray images to a content-hash-tagged, WebDataset-consumable JSONL manifest
and optional `(split, label)`-organized tar shards.

**Hard dependency:** `radiologist-utils` (`ImageReader`, `read_image`).

### Input layout contract

```
dataset/
├── images/
│   ├── NORMAL/    img001.png …
│   └── ABNORMAL/  img002.png …
└── masks/         ← optional sibling; must mirror images/ tree
    ├── NORMAL/    img001.png …
    └── ABNORMAL/  img002.png …
```

Label from parent directory name (ImageFolder convention).
Images must be exactly one directory deep inside their class folder.
Mask presence auto-detected — no second reader, no alignment risk.

---

### Module layout

```
radiologist-etl/src/radiologist/etl/
├── __init__.py
├── stats.py        # StatExtractor protocol, make_haralick() factory, lung_asymmetry()
├── ops.py          # lung_out_of_frame() — pure predicate, called in stats pass
├── filters.py      # filter_iqr(), filter_lung_out_of_frame()
├── split.py        # deterministic hash-based split assignment
├── processors.py   # StatsProcessor — ProcessPoolExecutor, mask resolution
├── manifest.py     # ManifestRecord, ParquetWriter, JsonlWriter
├── shards.py       # WebDataset .tar shard builder (optional step)
└── pipeline.py     # Prefect @task/@flow + Hydra @hydra.main entry point
```

---

### Artifact naming

Every ETL run computes a **run ID** — a 16-character SHA-256 prefix derived from the
run configuration and a lightweight data fingerprint (file count + total bytes of the
source directory). Identical config on unchanged data produces the same run ID, making
the pipeline idempotent: if `manifest-{run_id}.jsonl` already exists, the run can be
skipped. A `run_label` override in the Hydra config forces a new ID when you deliberately
want a fresh artifact despite identical inputs (e.g. A/B filter comparison).

```python
def compute_run_id(cfg: DictConfig, source: str, storage_options: Optional[dict] = None) -> str:
    # SHA-256[:16] of (sorted config dict + data fingerprint)
    # data fingerprint = file count + total bytes of source  — cheap, no image reads
    # cfg.run_label overrides the hash to force a new ID with identical inputs
    ...
```

```
artifact_dir/
├── stats-{run_id}.parquet           # raw stats, all images
├── stats-{run_id}-filtered.parquet  # after IQR + lung_out_of_frame filter
├── stats-{run_id}-split.parquet     # after split assignment
└── manifest-{run_id}.jsonl          # final, WebDataset-consumable

shard_root/                          # shards carry no run ID — content-stable,
├── train/                           # reusable across manifest versions
│   ├── NORMAL/   train-normal-000000.tar …
│   └── ABNORMAL/ train-abnormal-000000.tar …
├── val/
│   ├── NORMAL/   val-normal-000000.tar …
│   └── ABNORMAL/ val-abnormal-000000.tar …
└── test/
    ├── NORMAL/   …
    └── ABNORMAL/ …
```

Shards carry no run ID because they are content-stable derived artifacts.
The manifest is the versioned artifact; the `shard` field on each record ties lineage back to it.

---

### Data flow

```
run_id = compute_run_id(cfg, source)       SHA-256[:16] of config + data fingerprint
                                           run_label override forces a distinct ID

ImageReader(source)                        images only — no mask reader
        │
        ▼
StatsProcessor(masks_root=...)             ProcessPoolExecutor — N workers
Per image worker:
  ├─ read_image(image_path)
  ├─ _resolve_mask(image_path, images_root, masks_root)
  │    rel       = image_path[len(images_root):].lstrip("/")
  │    mask_path = masks_root.rstrip("/") + "/" + rel
  │    returns None on FileNotFoundError or masks_root=None
  ├─ haralick_extractor(image)             → dict[str, float]  (returned by make_haralick(...))
  ├─ lung_asymmetry(image, mask)           → dict[str, float], {} if mask=None
  └─ lung_out_of_frame(mask)              → bool, None if mask=None
        │
        ▼
stats-{run_id}.parquet                     per-image rows, flattened stat columns
        │
manifest filters                           no image I/O — Parquet DataFrame only
├─ filter_iqr(columns, factor)             sets excluded=True, reason="iqr:<col>"
└─ filter_lung_out_of_frame()              sets excluded=True, reason="lung_out_of_frame"
        │                                  no-op if column absent (no masks provided)
stats-{run_id}-filtered.parquet
        │
split assignment                           hash(filename) → train/val/test
        │                                  all records assigned, including excluded ones
stats-{run_id}-split.parquet
        │
        ▼
manifest-{run_id}.jsonl                    final — all records, excluded flagged
        │                                  manifest_id field on every record
[optional]
shards.py                                  reads manifest, skips excluded=True
→ {split}/{label}/{split}-{label}-{index:06d}.tar
→ manifest-{run_id}.jsonl updated with shard field per record
→ split-report-{run_id}.json written with observed vs configured ratios
```

---

### JSONL record schema

Every line in `manifest-{run_id}.jsonl` is a flat JSON object:

```json
{
  "manifest_id": "a3f9b2c1d4e5f6a7",
  "path": "gs://bucket/dataset/images/NORMAL/img001.png",
  "filename": "img001.png",
  "label": "NORMAL",
  "split": "train",
  "shard": "train/NORMAL/train-normal-000042.tar",
  "haralick_{feature}": 0.42,
  "asymmetry_ratio": 1.02,
  "asymmetry_diff": 0.03,
  "lung_out_of_frame": false,
  "excluded": false,
  "exclusion_reason": ""
}
```

- `manifest_id` is the run ID (same value as in the artifact filename); enables joins across manifests.
- `shard` is `null` until `shards.py` runs. Consumers check for non-null before loading shards.
- Stats are inlined at the top level (not nested under `"stats"`).

### Split report schema

`split-report-{run_id}.json` written by `build_shards`:

```json
{
  "run_id": "a3f9b2c1d4e5f6a7",
  "configured_ratios": {"train": 0.70, "val": 0.15, "test": 0.15},
  "observed": {
    "NORMAL":   {"train": 0.68, "val": 0.16, "test": 0.16, "excluded": 0.08},
    "ABNORMAL": {"train": 0.71, "val": 0.14, "test": 0.15, "excluded": 0.12}
  }
}
```
---

#### `stats.py`

```python
class StatExtractor(Protocol):
    def __call__(
        self,
        image: np.ndarray,
        metadata: dict[str, str],   # unused by current extractors — Protocol compliance
        mask: Optional[np.ndarray] = None,
    ) -> dict[str, float]: ...

HARALICK_PROPERTIES: frozenset[str] = frozenset({
    "mean", "std", "entropy", "contrast", "dissimilarity", "homogeneity", "energy", "correlation", "ASM"
})

def make_haralick(
    features: Optional[List[str]] = None,
    distances: Optional[List[int]] = None,
    angles: Optional[List[float]] = None,
) -> StatExtractor:
    # features:  subset of HARALICK_PROPERTIES; defaults to all nine; raises ValueError for unknown names
    # distances: pixel pair offsets, e.g. [1, 2, 3]; defaults to [1]
    # angles:    in radians, e.g. [0, pi/4, pi/2, 3*pi/4]; defaults to all four
    # Output keys: haralick_{feature} — one finite scalar per selected feature
    ...

# lung_asymmetry satisfies StatExtractor directly — no factory needed, nothing to configure
def lung_asymmetry(
    image: np.ndarray,
    metadata: dict[str, str],
    mask: Optional[np.ndarray] = None,
) -> dict[str, float]:
    # returns {} if mask is None
    # keys: asymmetry_ratio
```

#### `ops.py`

```python
def lung_out_of_frame(mask: np.ndarray) -> bool:
    # convert mask in grayscale if not
    # True if any nonzero pixel touches the image border
    # pure — no I/O, called inside each worker during stats pass
```

#### `filters.py`

```python
def filter_iqr(
    df: pd.DataFrame,
    columns: List[str],
    factor: float = 1.5,
) -> pd.DataFrame:
    # sets excluded=True, exclusion_reason="iqr:<col>"
    # multiple columns: reasons joined by "|"
    # never drops rows

def filter_lung_out_of_frame(df: pd.DataFrame) -> pd.DataFrame:
    # reads lung_out_of_frame column
    # sets excluded=True, exclusion_reason="lung_out_of_frame"
    # no-op if lung_out_of_frame column absent
```

#### `split.py`

```python
def assign_split(filename: str, ratios: dict[str, float]) -> str:
    # raises ValueError if sum(ratios.values()) != 1.0 (float tolerance)
    # deterministic: same filename always maps to the same split
    ...
```

#### `manifest.py`

```python
@dataclass
class ManifestRecord:
    manifest_id: str                        # run ID, same for all records in a run
    path: str
    filename: str
    label: str
    split: str
    stats: dict[str, float]                 # flattened at write time
    lung_out_of_frame: Optional[bool]       # None when masks_root=None
    excluded: bool = False
    exclusion_reason: str = ""
    shard: Optional[str] = None             # set by shards.py, null until then

class ParquetWriter:
    def write(
        self,
        records: List[ManifestRecord],
        destination: str,
        storage_options: Optional[dict] = None,
    ) -> None:
        # fsspec-backed — local or remote
        # flattens stats dict into columns
        # lung_out_of_frame as nullable boolean column
        # shard as nullable string column
        ...

class JsonlWriter:
    def write(
        self,
        records: List[ManifestRecord],
        destination: str,
        storage_options: Optional[dict] = None,
    ) -> None:
        # fsspec-backed — local or remote
        # stats inlined at top level (not nested)
        # all records written regardless of excluded flag — consumers filter
        ...
```

#### `processors.py`

```python
class StatsProcessor:
    def __init__(
        self,
        extractors: List[StatExtractor],
        workers: int = os.cpu_count(),
    ) -> None: ...

    def run(
        self,
        reader: BaseImageReader,
        manifest_id: str,
        masks_root: Optional[str] = None,
    ) -> List[ManifestRecord]:
        # images_root = reader.source — mask path derived by string prefix stripping (not Path.relative_to)
        # per-image errors logged and skipped — do not abort the run
        # submits _process_one per image via ProcessPoolExecutor
        # progress via rich progress bar
        ...

# top-level module function — required by ProcessPoolExecutor for multiprocessing
def _process_one(
    image_path: str,
    images_root: str,
    masks_root: Optional[str],
    manifest_id: str,
    extractors: List[StatExtractor],
    storage_options: Optional[dict],
) -> ManifestRecord:
    # Calls read_image and _resolve_mask


def _resolve_mask(
    image_path: str,
    images_root: str,
    masks_root: Optional[str],
    storage_options: Optional[dict],
) -> Optional[np.ndarray]:
    # works for local and remote URIs — plain string prefix stripping,
    # not Path.relative_to() which mangles URI schemes
```

#### `shards.py`

```python
def build_shards(
    manifest_path: str,
    shard_root: str,
    ratios: dict[dict, float],  # same as in assign_split
    shard_size: int = 1000,
    start_shard_index: Optional[dict[tuple[str, str], int]] = None,
    storage_options: Optional[dict] = None,
) -> str:
    # reads manifest-{uuid}.jsonl
    # skips excluded=True records
    # groups by (split, label)
    # writes {shard_root}/{split}/{label}/{split}-{label}-{index:06d}.tar
    # each tar sample: <stem>.png (raw bytes via fsspec) + <stem>.cls (label UTF-8)
    # start_shard_index: per (split, label) offset for incremental runs, default 0
    # patches shard field in each manifest record; rewrites manifest in place
    # logs observed vs configured split ratios per label — caller should inspect when
    #   exclusion rate is high (>15%) as effective ratios among included records may drift
    #  writes a split-report-{run_id}.json alongside the manifest
    # returns updated manifest path
```

**Split-before-filter design note:**

Split assignment runs on all records, including excluded ones. This is intentional:
hash-based assignment is per-record independent — excluding some records does not affect
any other record's split. Effective ratio drift is proportional to the exclusion rate
(small for typical IQR outlier removal, ≈5–10%). The benefit is reproducibility: if filter
parameters change in a later run and some previously excluded records are reactivated, they
recover their original split assignment without reshuffling the included records.

`build_shards` logs the observed vs configured ratios per `(split, label)` group. If drift
exceeds an acceptable threshold, the caller should re-run with a new `run_label` and adjusted
filter parameters rather than rebalancing shards post-hoc.

#### `pipeline.py`

Two layers per stage — portable core and Prefect shell:

- **Portable core** (`_` prefix): pure business logic, no Prefect imports. Reusable as  a `DoFn` for Beam support.
- **Prefect shell** (`@task`): thin wrapper — calls core, creates artifact. `create_link_artifact` / `create_table_artifact` live in the shell only, never in the core.

```python
def compute_run_id(cfg: DictConfig, source: str, storage_options: Optional[dict] = None) -> str:
    # SHA-256[:16] of (sorted config dict + data fingerprint)
    # data fingerprint = file count + total bytes of source — cheap, no image reads
    # cfg.run_label overrides the hash to force a new ID with identical inputs
    ...

# ── Portable cores ────────────────────────────────────────────────────────────

def _compute_stats(
    source: str,
    run_id: str,
    masks_root: Optional[str],
    extractors: List[StatExtractor],
    workers: int,
    artifact_dir: str,
    storage_options: Optional[dict] = None,
) -> str: ...  # → stats-{run_id}.parquet

def _apply_filters(
    parquet_path: str,
    iqr_columns: List[str],
    factor: float = 1.5,
) -> str: ...  # → stats-{run_id}-filtered.parquet

def _assign_splits(
    parquet_path: str,
    ratios: dict[str, float],
) -> str: ...  # → stats-{run_id}-split.parquet

def _write_jsonl(
    parquet_path: str,
    destination: str,
    storage_options: Optional[dict] = None,
) -> str: ...  # → manifest-{run_id}.jsonl

def _build_shards(
    manifest_path: str,
    shard_root: str,
    ratios: dict[str, float],
    shard_size: int = 1000,
    start_shard_index: Optional[dict[tuple[str, str], int]] = None,
) -> str: ...  # → updated manifest-{run_id}.jsonl + split-report-{run_id}.json

# ── Prefect shells ────────────────────────────────────────────────────────────

@task
def compute_stats(...) -> str:
    path = _compute_stats(...)
    create_link_artifact(key=f"stats-{run_id}", link=path, ...)
    return path

@task
def apply_filters(...) -> str:
    path = _apply_filters(...)
    create_link_artifact(key=f"stats-{run_id}-filtered", link=path, ...)
    return path

@task
def assign_splits(...) -> str:
    path = _assign_splits(...)
    create_link_artifact(key=f"stats-{run_id}-split", link=path, ...)
    return path

@task
def write_jsonl(...) -> str:
    path = _write_jsonl(...)
    create_link_artifact(key=f"manifest-{run_id}", link=path, ...)
    return path

@task
def build_shards_task(...) -> str:
    path = _build_shards(...)
    create_table_artifact(key=f"split-report-{run_id}", ...)
    return path

# ── Flow ──────────────────────────────────────────────────────────────────────

@flow
def etl_flow(cfg: DictConfig) -> str:
    # compute_stats → apply_filters → assign_splits → write_jsonl → (optional) build_shards_task
    # each task uses cache_key_fn=task_input_hash + result_storage for idempotent crash recovery
    # bypass parameters (resume_from_parquet, etc.) allow entering at any stage
    ...

@hydra.main(config_path="conf", config_name="etl")
def main(cfg: DictConfig) -> None:
    etl_flow(cfg)
```

Each prefect task uses cache_key_fn=task_input_hash + result_storage for idempotent crash recovery. The flow accepts bypass parameters to enter at any stage from a prior run.

Each prefect task also creates a Prefect artifact (prefect.artifacts) for UI visibility and lineage — separate from the cached result. Tasks return path strings; artifacts are a side effect.
| Task | Artifact key | Type |
|---|---|---|
| `compute_stats` | `stats-{run_id}` | `LinkArtifact` — path + total record count |
| `apply_filters` | `stats-{run_id}-filtered` | `LinkArtifact` — path + excluded count |
| `assign_splits` | `stats-{run_id}-split` | `LinkArtifact` — path + per-split counts |
| `write_jsonl` | `manifest-{run_id}` | `LinkArtifact` — path |
| `build_shards_task` | `split-report-{run_id}` | `TableArtifact` — observed vs configured ratios |

Hydra config keys: source, masks_root, destination, artifact_dir, iqr_columns, iqr_factor, split_ratios, workers, storage_options, build_shards, shard_root, shard_size, run_label, resume_from_parquet, resume_from_filtered, resume_from_split, resume_from_manifest, haralick.features, haralick.distances, haralick.angles.

**How run ID and task caching work together:**

- `compute_run_id` hashes config + data fingerprint; same inputs → same run ID → artifacts already exist → pipeline can skip entirely.
- `run_label` overrides the hash to force new artifacts when deliberately re-running with identical inputs.
- `cache_key_fn=task_input_hash` — Prefect hashes all task inputs; identical inputs return the cached path string without re-running the task body.
- `result_storage` — persists the cached return value across process restarts, not just within a single run.
- On a crash-and-resume, completed tasks are skipped; the flow picks up from the first uncached task.
- On an incremental run with new images, pass `resume_from_parquet=<path>` to skip `compute_stats` and re-run only downstream stages.

---

### WebDataset consumption (radiologist-core)

Shards are organized by `(split, label)` so consumers need no filtering logic:

```python
class ChestXRayDataModule(LightningDataModule):

    def _class_dataset(self, split: str, label: str) -> wds.WebDataset:
        pattern = f"{self.shard_root}/{split}/{label}/*.tar"
        return (
            wds.WebDataset(wds.ResampledShards(pattern))
            .decode("pil")
            .to_tuple("png", "cls")
            .shuffle(self.shuffle_buffer)
            .map_tuple(self.transform, int)
        )

    def train_dataloader(self) -> DataLoader:
        datasets = [self._class_dataset("train", lbl) for lbl in self.labels]
        # wds.RandomMix balances classes at consumption time
        mixed = wds.RandomMix(datasets, probs=self.class_weights)
        return DataLoader(mixed, batch_size=self.batch_size, num_workers=self.workers)

    def val_dataloader(self) -> DataLoader:
        ds = wds.WebDataset(f"{self.shard_root}/val/**/*.tar")
            .decode("pil").to_tuple("png", "cls")
        return DataLoader(ds, batch_size=self.batch_size, num_workers=self.workers)
```

`ResampledShards` samples with replacement — minority class loops rather than being exhausted.
Transforms are stateless callables; thread-safe by construction with `torchvision` and Albumentations.

---

### Issue specs

See the table below for all 8 issues formatted with full context,
user stories, acceptance criteria, technical notes, and dependency graph.
